r/src/csv.cpp

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow_types.h"
+
+using namespace Rcpp;
+
+// [[Rcpp::export]]
+std::shared_ptr<arrow::csv::ReadOptions> csv___ReadOptions__initialize(List_ options) {
+  auto res =
+      std::make_shared<arrow::csv::ReadOptions>(arrow::csv::ReadOptions::Defaults());
+  res->use_threads = options["use_threads"];
+  res->block_size = options["block_size"];
```

Comment:
How does List_ deal with type coercion?

Reply:
It's taken care of, `List_::operator[]` gives you a proxy object that then goes through `Rcpp::as< {the type you want}>(.)`. 

There's no support for a single `char` so I had to add this : 

```cpp
inline char get_char(SEXP x) { return CHAR(STRING_ELT(x, 0))[0]; }
```

Reply:
Modified it is that we'd get an Rcpp error instead of an R error if for some reason something else than a character vector is given : 

```r
inline char get_char(CharacterVector x) { return CHAR(STRING_ELT(x, 0))[0]; }
```